File: web/src/api/confluence.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export type Market = 'CN' | 'US';
export type SignalKey = 'industry' | 'trend' | 'quant' | 'etf' | 'dragonTiger';
export interface ConfluenceSignal {
  sourceModule: string; status: 'positive' | 'neutral' | 'negative' | 'unavailable';
  weight: number; score: number | null; tradeDate: string | null; sourceGeneratedAt: string | null;
  evidence: Record<string, unknown>; reasons: string[];
}
export interface ConfluenceItem {
  instrumentId: number; code: string; name: string; confluenceScore: number | null;
  availableWeight: number; availableSignalCount: number; positiveSignalCount: number;
  eligible: boolean; strongConfluence: boolean; signals: Record<SignalKey, ConfluenceSignal>;
  reasons: string[]; generatedAt: string;
}
export interface ConfluenceRanking {
  market: Market; tradeDate: string | null; generatedAt: string | null; algorithmVersion: string | null;
  sourceAvailability: Record<SignalKey, { status: string; count: number; tradeDate: string | null; reason?: string }>;
  summary: { total: number; eligible: number; strongConfluence: number; ignitionIndustryStrong: number; topIndustryConfluence: number };
  rules: { minSignals: number; strongMinSignals: number; strongMinPositive: number; strongMinScore: number };
  total: number; items: ConfluenceItem[];
}
export interface RankingParams {
  market: Market; trade_date?: string; min_score?: number; min_signals?: number; industry?: string;
  lifecycle?: string; early_only?: boolean; top_industry?: boolean; strong_only?: boolean; limit?: number;
}
const base = '/api/v1/confluence';
export const confluenceApi = {
  async ranking(params: RankingParams): Promise<ConfluenceRanking> {
    return toCamelCase((await apiClient.get(`${base}/ranking`, { params })).data);
  },
  async dates(market: Market): Promise<string[]> { return (await apiClient.get(`${base}/dates`, { params: { market } })).data; },
  async detail(code: string, market: Market, tradeDate: string): Promise<{ market: Market; tradeDate: string; algorithmVersion: string; item: ConfluenceItem }> {
    return toCamelCase((await apiClient.get(`${base}/${encodeURIComponent(code)}`, { params: { market, trade_date: tradeDate } })).data);
  },
  async run(market: Market, tradeDate?: string): Promise<{ taskId: string }> {
    return toCamelCase((await apiClient.post(`${base}/run`, { market, trade_date: tradeDate })).data);
  },
};

File: src/finance_analysis/interfaces/api/v1/schemas/confluence.py
"""Explicit read and asynchronous run contracts."""

from datetime import date, datetime
from typing import Any, Literal
from pydantic import BaseModel, Field, model_validator

Market = Literal["CN", "US"]


class Signal(BaseModel):
    source_module: str
    status: Literal["positive", "neutral", "negative", "unavailable"]
    weight: int
    score: float | None
    trade_date: date | None
    source_generated_at: datetime | None
    evidence: dict[str, Any]
    reasons: list[str]


class ConfluenceItem(BaseModel):
    instrument_id: int
    code: str
    name: str
    confluence_score: float | None
    available_weight: int
    available_signal_count: int
    positive_signal_count: int
    eligible: bool
    strong_confluence: bool
    signals: dict[str, Signal]
    reasons: list[str]
    generated_at: datetime


class Ranking(BaseModel):
    market: Market
    trade_date: date | None
    generated_at: datetime | None
    algorithm_version: str | None
    source_availability: dict[str, Any]
    summary: dict[str, int]
    total: int
    items: list[ConfluenceItem]


class Detail(BaseModel):
    market: Market
    trade_date: date
    algorithm_version: str
    item: ConfluenceItem


class RunRequest(BaseModel):
    market: Market = "CN"
    trade_date: date | None = Field(default=None)

    @model_validator(mode="after")
    def no_future(self):
        from finance_analysis.market_review.trading_calendar import get_market_now

        if self.trade_date and self.trade_date > get_market_now(self.market.lower()).date():
            raise ValueError("Cannot generate a future confluence date")
        return self
